### PR TITLE
FIPO loss

### PR DESCRIPTION
**Pull request by kdubovikov**
# What does this PR do?
This is a port of https://github.com/qwenpilot/FIPO as a GRPOTrainer loss function. 

Additional manual validation outside the committed test suite:
- ran short GRPO/FIPO training jobs on `AI-MO/NuminaMath-TIR`
- confirmed that both trainers produce nonzero updates when reward variance is present
- confirmed that with `num_iterations > 1`, FIPO’s inner reuse steps show nonzero `log_ratio`, nonzero Future-KL, and influence weights that move away from `1.0`, indicating that the FIPO-specific reweighting path is active

## Before submitting

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link to it if that's the case.
- [X] Did you make sure to update the documentation with your changes?
- [X] Did you write any new necessary tests?

## AI writing disclosure

We welcome the use of AI tools to help with contributions. For transparency and to help us improve our review process, please indicate the level of AI involvement in this PR.

- [ ] No AI usage: the PR was written entirely by a human.
- [X] AI-assisted: some parts were suggested or improved by AI, but the PR was written and reviewed by a human.
- [ ] AI-generated: the PR was mostly or fully generated by an AI tool.

---

> [!NOTE]
> **Medium Risk**
> Adds a new GRPO loss formulation that changes the core token-loss computation and masking behavior during training, which could affect stability/metrics for users opting into `loss_type="fipo"`. Risk is contained to the new loss path plus shared loss-aggregation code touched to support a per-loss `loss_mask`.
> 
> **Overview**
> Adds **FIPO (Future-KL Influenced Policy Optimization)** as a new `GRPOTrainer` `loss_type`, including Future-KL influence-weight computation, dual-clipping behavior for negative-advantage tokens, optional sequence dropping, and new `fipo/*` training metrics.
> 
> Extends `GRPOConfig` with FIPO-specific hyperparameters (decay, chunking, clip thresholds, influence clipping, safety threshold) plus validation/warnings (e.g., `importance_sampling_level` ignored for FIPO, `use_liger_kernel` unsupported).
> 
> Updates docs with a FIPO recipe snippet and expands the GRPO loss-type test matrix to cover `loss_type="fipo"`.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 20463989e8d9fef8dbba7230c7fb11bdd7f6387f. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: trl/trainer/grpo_trainer.py [hunk 1]
```diff
@@ -2255,6 +2263,92 @@ def get_gamma_weights(
...
+        pg_losses2 = -weighted_advantages * torch.clamp(ratio, 1 - epsilon_low, 1 + epsilon_high)
+        clip_pg_losses1 = torch.maximum(pg_losses1, pg_losses2)
+        pg_losses3 = -weighted_advantages * clip_ratio_c
+        clip_pg_losses2 = torch.minimum(pg_losses3, clip_pg_losses1)
+
+        lower_clip_mask = (advantages < 0) & (clip_pg_losses1 > pg_losses3) & mask_bool
+        low_clip_token_counts = lower_clip_mask.sum(dim=1)
+        seq_valid_mask = (low_clip_token_counts <= 1).unsqueeze(1)
+        loss_mask = (mask_bool & seq_valid_mask).to(dtype=log_ratio_per_token.dtype)
+        per_token_loss = torch.where(weighted_advantages < 0, clip_pg_losses2, clip_pg_losses1)
```
[the remark this answers is not in the document]
Comment (kdubovikov) [reply trl/trainer/grpo_trainer.py]: fixed
